buildkite-agent-scaler/main.tf

```diff
@@ -0,0 +1,107 @@
+data aws_region current {}
+
+locals {
+  aws_region        = data.aws_region.current.name
+  default_s3_bucket = local.aws_region == "us-east-1" ? "buildkite-lambdas" : "buildkite-lambdas-${local.aws_region}"
+  s3_bucket         = var.s3_bucket == "" ? local.default_s3_bucket : var.s3_bucket
+  function_name     = var.function_name
+}
+
+resource aws_iam_role this {
+  name_prefix = format("%s-", local.function_name)
+
+  assume_role_policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Action": "sts:AssumeRole",
+      "Principal": {
+        "Service": "lambda.amazonaws.com"
+      },
+      "Effect": "Allow",
+      "Sid": ""
+    }
+  ]
+}
+EOF
+}
+
+resource aws_iam_role_policy lambda_autoscaling {
+  name = "lambda_autoscaling"
+  role = aws_iam_role.this.id
+
+  policy = <<-EOF
+  {
+    "Version": "2012-10-17",
+    "Statement": [
+      {
+        "Action": [
+          "autoscaling:SetDesiredCapacity",
+          "autoscaling:DescribeAutoScalingGroups",
+          "cloudwatch:PutMetricData",
+          "logs:CreateLogStream",
+          "logs:PutLogEvents"
+        ],
+        "Effect": "Allow",
+        "Resource": "*"
+      }
+    ]
+  }
+  EOF
+}
+
+resource aws_lambda_function this {
```

Comment:
I don't know if you want to use an open source module for this, but this might save some trouble.

https://github.com/claranet/terraform-aws-lambda